src/util/coll_selection/coll_selection_tree_from_cvars.h

```diff
@@ -0,0 +1,346 @@
+
+    coll = MPIU_COLL_SELECTION_create_node(parent, MPIU_COLL_SELECTION_COLLECTIVE,
+                                           MPIU_COLL_SELECTION_CONTAINER, coll_id, 1);
+
+    algo = MPIU_COLL_SELECTION_create_leaf(coll, MPIU_COLL_SELECTION_CONTAINER, 1, (void *)
+                                           MPIDI_Coll_intra_composition_default_cnt[coll_id]);
+    return coll;
+}
+
+static MPIU_COLL_SELECTION_storage_handler
+MPIU_COLL_SELECTION_create_coll_tree_net_default(MPIU_COLL_SELECTION_storage_handler parent,
+                                                 int coll_id)
+{
+    MPIU_COLL_SELECTION_storage_handler coll = MPIU_COLL_SELECTION_NULL_ENTRY;
+    MPIU_COLL_SELECTION_storage_handler algo = MPIU_COLL_SELECTION_NULL_ENTRY;
+
+    coll = MPIU_COLL_SELECTION_create_node(parent, MPIU_COLL_SELECTION_COLLECTIVE,
+                                           MPIU_COLL_SELECTION_CONTAINER, coll_id, 1);
+
+    algo = MPIU_COLL_SELECTION_create_leaf(coll, MPIU_COLL_SELECTION_CONTAINER, 1,
+                                           (void *) MPIDI_OFI_Coll_default_cnt[coll_id]);
+
+    return coll;
+}
+
+static MPIU_COLL_SELECTION_storage_handler
+MPIU_COLL_SELECTION_create_coll_tree_shm_default(MPIU_COLL_SELECTION_storage_handler parent,
+                                                 int coll_id)
+{
+    MPIU_COLL_SELECTION_storage_handler coll = MPIU_COLL_SELECTION_NULL_ENTRY;
+    MPIU_COLL_SELECTION_storage_handler algo = MPIU_COLL_SELECTION_NULL_ENTRY;
+
+    coll = MPIU_COLL_SELECTION_create_node(parent, MPIU_COLL_SELECTION_COLLECTIVE,
+                                           MPIU_COLL_SELECTION_CONTAINER, coll_id, 1);
+
+    algo = MPIU_COLL_SELECTION_create_leaf(coll, MPIU_COLL_SELECTION_CONTAINER, 1,
+                                           (void *) MPIDI_POSIX_Coll_default_cnt[coll_id]);
+    return coll;
+}
+
+static MPIU_COLL_SELECTION_storage_handler
+MPIU_COLL_SELECTION_create_bcast_tree_inter_compositions(MPIU_COLL_SELECTION_storage_handler parent,
+                                                         int coll_id)
+{
+    MPIU_COLL_SELECTION_storage_handler bcast = MPIU_COLL_SELECTION_NULL_ENTRY;
+    MPIU_COLL_SELECTION_storage_handler algo = MPIU_COLL_SELECTION_NULL_ENTRY;
+
+    bcast = MPIU_COLL_SELECTION_create_node(parent, MPIU_COLL_SELECTION_COLLECTIVE,
+                                            MPIU_COLL_SELECTION_CONTAINER, coll_id, 1);
+
+    algo = MPIU_COLL_SELECTION_create_leaf(bcast, MPIU_COLL_SELECTION_CONTAINER, 1,
+                                           (void *) &MPIDI_Bcast_inter_composition_alpha_cnt);
+
+    return bcast;
+}
+
+static MPIU_COLL_SELECTION_storage_handler
+MPIU_COLL_SELECTION_create_bcast_tree_topo_aware_compositions(MPIU_COLL_SELECTION_storage_handler
+                                                              parent, int coll_id)
+{
+/*  The tree which creates below stands for the following logic:
```

Comment:
This where this PR loses me. I can't see a reason to include this in open-source MPICH unless there is greater interest from downstream. This relatively readable conditional is replaced by an abstract tree structure that will require a lot of effort to understand, validate, and/or modify.

Reply:
At minimum, I'd like to hear from someone at Cray if they are interested in this. I'm not sure if our regular ch4 collaborators are on github, so I'll reach out by email.